### PR TITLE
Support video files in public background

### LINKED ISSUE #21289: Allow video files for "public background" field
You can currently set a custom background image, but it'd be really nice if you could set a MP4. The field itself already accepts it, but it tries rendering the video file as a background-image. The only workaround is to use an unoptimized GIF.

### PR DESCRIPTION
**Pull request by rijkvanzanten**
## Scope

What's changed:

- When a video type (like MP4) is uploaded as the public_background, render it as an autoplaying looping muted `<video>`
- Fixed a bug (I think) in the assets endpoint

## Potential Risks / Drawbacks

- It's a small breaking change in the output of the server info endpoint, as it now returns the file info rather than just the ID string

## Review Notes / Questions

- The assets endpoint has an explicit permissions check, but then also had regular permissions enabled for the file service it uses. It caused the video to be forbidden to be loaded for non-authenticated users. The assets endpoint already has a strict permissions check for file read, so I feel like that was redundant/an oversight. 

---

Fixes directus/directus#21289

### CODE REVIEW

File: api/src/services/assets.ts [hunk 1]
```diff
@@ -38,8 +38,8 @@ export class AssetsService {
 	constructor(options: AbstractServiceOptions) {
 		this.knex = options.knex || getDatabase();
 		this.accountability = options.accountability || null;
-		this.filesService = new FilesService(options);
 		this.authorizationService = new AuthorizationService(options);
+		this.filesService = new FilesService({ ...options, accountability: undefined });
```
Comment (rijkvanzanten) [inline api/src/services/assets.ts:42]: This is the potentially risky change, but the auth check on line 69 should mean this file service doesn't have to be authenticated

File: app/src/stores/server.ts [hunk 2]
```diff
@@ -28,7 +28,7 @@ export type Info = {
 		theme_light_overrides: Record<string, unknown> | null;
 		theme_dark_overrides: Record<string, unknown> | null;
 		public_foreground: string | null;
-		public_background: string | null;
+		public_background: { id: string; type: string } | null;
```
Comment (DanielBiegler) [inline app/src/stores/server.ts:31]: As of right now we gotta update changes to server infos both here and the sdk, else theyre out of sync, see:

https://github.com/directus/directus/blob/5153f53574d86ad6e3768280a1c86459c20651dc/sdk/src/schema/settings.ts#L13-L15

**Review by DanielBiegler — DISMISSED** [review]
Love the idea of supporting videos, nice!

A small thing: The public foreground is broken when using the video background, see:

Screenshot from 2024-02-05 13-56-59

It should be centered on top like when using an image

image

Should be a short css fix I presume :smile: :+1:

**Review by br41nslug — CHANGES_REQUESTED** [review]
The docs for this pr should make mention of what to look out for here:
- Not all video formats/encodings can be streamed by browser resulting in the video needing to be downloaded fully before playing.
- Both supported and the optimal formats/settings can vary per browser.
- Mobile browsers will start pre-loading (or even fully loading if streaming is not supported) the entire video even tho it is never shown
- ~~Permissions need to be set to public same as for an image~~ The assets endpoint has an override that always allows assets to be accessed that are used in the public page settings

**Comment by rijkvanzanten** [comment]
@br41nslug Regarding points 1, 2, 3: the current alternative that we've seen people use is using GIFs as a workaround, which is exponentially worse for all the same reasons 😅 

> Permissions need to be set to public same as for an image

This is inaccurate! The assets endpoint has an override that always allows assets to be accessed that are used in the public page settings 👍🏻

**Comment by br41nslug** [comment]
>  the current alternative that we've seen people use is using GIFs as a workaround, which is exponentially worse for all the same reasons 😅

True but this feature lowers the bar by a lot, people wont be encoding a 10Gig GIF but might accidentally select a 4K multi-gig video.

> This is inaccurate! The assets endpoint has an override that always allows assets to be accessed that are used in the public page settings 👍🏻

Interesting, is that documented somewhere? 👀

**Comment by rijkvanzanten** [comment]
> To fit the screen space though those gifs are going to have to be mighty massive to look decent right?

Haha yes which is what people are doing now 🙃 This PR is really just to enable that same type of animated background functionality in a way that doesn't require dozens of gigs worth of GIF 🙂

**Comment by DanielBiegler** [comment]
> I would propose a size limit

I dont like dictating how users should use their own instance for such a subjective thing. Maybe they do want to have a 4k 1gig video of a sunset, its ultimately their choice. We should make it good for mobile though so we dont waste mobile data if the video is not visible, thats less of a choice and more optimization. :+1:

**Comment by M-GRV** [comment]
> > I would propose a size limit
> 
> I dont like dictating how users should use their own instance for such a subjective thing. Maybe they do want to have a 4k 1gig video of a sunset, its ultimately their choice. We should make it good for mobile though so we dont waste mobile data if the video is not visible, thats less of a choice and more optimization. 👍

I meant something that could be imposed by the admin. 

But I agree. In any case at minimum possibly a subtext warning of a file size that says hey this might not work as expected due to how big the file is? I'm just looking at the inevitable support requests saying "my video background doesn't work." thats all.

**Comment by br41nslug** [comment]
> I know mp4/webm support on modern browsers have come a long way.

@M-GRV 

Correct but thats where video gets complicated, there is no universal "mp4" it is just a container which can hold encodings like (not an exhaustive list):
- MPEG-4 (very well supported)
- h264 (very well supported) but the 10bit variant not so much
- h265 (very limited support)
- AV1 (limited support)

All these codecs (and others not mentioned) have a multitude of configuration options that affect everything from quality, size to web streamability.

https://en.wikipedia.org/wiki/HTML5_video#Browser_support
https://videojs.com/html5-video-support/

**Comment by M-GRV** [comment]
> > I know mp4/webm support on modern browsers have come a long way.
> 
> @M-GRV
> 
> Correct but thats where video gets complicated, there is no universal "mp4" it is just a container which can hold encodings like (not an exhaustive list):
> 
> * MPEG-4 (very well supported)
> * h264 (very well supported) but the 10bit variant not so much
> * h265 (very limited support)
> * AV1 (limited support)
> 
> All these codecs (and others not mentioned) have a multitude of configuration options that affect everything from quality, size to web streamability.
> 
> https://en.wikipedia.org/wiki/HTML5_video#Browser_support https://videojs.com/html5-video-support/

This is why if I ever had to deal with video I just used Wowza to provide an HLS stream and just be done with it. (obviously not the case here) (I do know there are CDNs now that provide this part built-in by just uploading the video which is dope, but again not the case here)

But I guess the easiest thing is if video can play, play it. If not, fallback. :)  Also kinda off topic, excited that AV1 is finally making waves.

**Comment by rijkvanzanten** [comment]
@br41nslug Tracking the docs updates in linked issue #21378 🙂 With the other notes resolved, I believe this is ready to merge 👍🏻

File: api/src/services/assets.ts [hunk 3]
```diff
@@ -40,6 +40,7 @@ export class AssetsService {
 		this.accountability = options.accountability || null;
 		this.filesService = new FilesService({ ...options, accountability: null });
 		this.authorizationService = new AuthorizationService(options);
+		this.filesService = new FilesService({ ...options, accountability: undefined });
```
Comment (br41nslug) [inline api/src/services/assets.ts:43]:
```suggestion
```

Why are we setting `this.filesService` twice here? Did you mean to update line 41 or create a new FilesService under a different property name 🤔

**Review by br41nslug — CHANGES_REQUESTED** [review]

File: api/src/services/assets.ts (hunk 3 above)
Comment (rijkvanzanten) [reply api/src/services/assets.ts]: Wrong merge conflict fix :)

**Review by br41nslug — APPROVED** [review]

**Pull request merged**